server/chat-plugins/friends.ts

```diff
@@ -0,0 +1,905 @@
+	checkLastLogin: `SELECT last_login FROM friends WHERE friend = ?`,
+	deleteLogin: `UPDATE friends SET last_login = null WHERE friend = ?`,
+	expire: (
+		`DELETE FROM friend_requests WHERE EXISTS` +
+		`(SELECT sent_at FROM friend_requests WHERE should_expire(sent_at) = 1)`
+	),
+	hidingLogin: `SELECT * FROM friend_settings WHERE name = ? AND send_login_data = 1`,
+	hideLogin: `REPLACE INTO friend_settings (name, send_login_data) VALUES (?, 1)`,
+	showLogin: `DELETE FROM friend_settings WHERE name = ? AND send_login_data = 1`,
+};
+
+
+const TRANSACTIONS: {[k: string]: (input: any[]) => DatabaseResult} = {
+	rename: requests => {
+		for (const request of requests) {
+			const [oldID, newID] = request;
+			const oldFriends = statements.get.all(oldID, MAX_FRIENDS).map(item => item.friend);
+			const newRequests = statements.getReceived.all(oldID);
+			const sentRequests = statements.getSent.all(oldID);
+			if (newRequests.length || sentRequests.length) {
+				throw new Chat.ErrorMessage(`You have pending friend requests, and so cannot transfer friends.`);
+			}
+			if (!oldFriends.length) {
+				throw new Chat.ErrorMessage(`You have no friends to transfer.`);
+			}
+			if (oldFriends.includes(newID)) {
+				// unfriend the two
+				statements.delete.run({userid: newID});
+			}
+			statements.renameFriend.run({oldID, newID});
+			statements.renameUserid.run({oldID, newID});
+			statements.rename.run(oldID, newID, Date.now());
+		}
+		return {result: []};
+	},
+	send: requests => {
+		for (const request of requests) {
+			const [senderID, receiverID] = request;
+			const sentRequests = statements.getSent.all(senderID).map(item => item.receiver);
+			const friends = statements.get.all(senderID, MAX_FRIENDS).map(item => item.friend);
+			if (friends.length >= MAX_FRIENDS) {
+				throw new Chat.ErrorMessage(`You are at the maximum number of friends.`);
+			}
+			const existingFriendship = statements.findFriendship.all({user1: senderID, user2: receiverID});
+			if (existingFriendship.length) {
+				throw new Chat.ErrorMessage(`You are already friends with '${receiverID}'.`);
+			}
+			if (sentRequests.includes(receiverID)) {
+				throw new Chat.ErrorMessage(`You have already sent a friend request to '${receiverID}'.`);
+			}
+			if (sentRequests.length >= MAX_REQUESTS) {
+				throw new Chat.ErrorMessage(
+					`You already have ${MAX_REQUESTS} outgoing friend requests. Use "/friends view sent" to see your outgoing requests.`
+				);
+			}
+			if (friends.includes(receiverID)) {
+				throw new Chat.ErrorMessage(`You have already friended '${receiverID}'.`);
+			}
+			statements.insertRequest.run(senderID, receiverID, Date.now());
+		}
+		return {result: []};
+	},
+	add: requests => {
+		for (const request of requests) {
+			const [senderID, receiverID] = request;
+			const last_login = Date.now();
+			statements.add.run({userid: senderID, friend: receiverID, last_login});
+			statements.add.run({friend: senderID, userid: receiverID, last_login});
```

Comment:
What is the point of storing this relationship twice? Can we adjust the schema/queries so only storing it once is necessary?